### PR TITLE
[query][smm 14] Memory management for TableMapPartitions

### PR DESCRIPTION
**Pull request by patrick-schultz**
~~Stacked on #9312.~~

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Compile.scala [hunk 1]
```diff
@@ -177,18 +181,24 @@ object CompileIterator {
...
+    val outerRegionField = fb.genFieldThisRef[Region]("outerRegion")
+    val eltRegionField = fb.genFieldThisRef[Region]("eltRegion")
+    val setF = fb.newEmitMethod("setRegions", FastIndexedSeq(CodeParamType(typeInfo[Region]), CodeParamType(typeInfo[Region])), CodeParamType(typeInfo[Unit]))
+    setF.emit(Code(outerRegionField := setF.getCodeParam[Region](1), eltRegionField := setF.getCodeParam[Region](2)))
+
     val stepF = fb.apply_method
     val stepFECB = stepF.ecb
 
-    val region = StagedRegion(EmitRegion.default(stepF).region, allowSubregions = false)
+    val eltRegion = StagedRegion(eltRegionField, allowSubregions = false)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/Compile.scala:192]: we'll flip these to `true` when we solve the mixed memory management design, right?

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 2]
```diff
@@ -1637,18 +1637,15 @@ class Emit[C](
               val updateMap = Map(fields: _*)
               val srvb = new StagedRegionValueBuilder(mb, x.pType, region.code)
 
-              val addFields = { (newMB: EmitMethodBuilder[C], t: PType, v: EmitCode) =>
-                Code(
-                  v.setup,
-                  v.m.mux(srvb.setMissing(), srvb.addIRIntermediate(t)(v.v)),
-                  srvb.advance())
-              }
+              def addFields(t: PType, v: EmitCode): Code[Unit] = Code(
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:1640]: why these changes? Did something break?

File: hail/src/main/scala/is/hail/expr/ir/Compile.scala (hunk 1 above)
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/Compile.scala:192]: I don't know yet what the final design will look like for setting and passing down the allocation strategy, but as things are now I think we'd want `outerRegion` to allow subregions, but not `eltRegion`.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala (hunk 2 above)
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:1640]: The `emitSelf.emit` call is bad, because it will use the Region argument of `mb` rather than the region passed to emit. This looked to me like some cruft left from the method splitter change, so I cleaned it up while I was fixing the bug. I don't remember anything breaking, I think I found this by looking at where the `emit` overload that uses the region argument is used.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:1640]: ahhhh, got it.
